torch/fx/experimental/proxy_tensor.py

```diff
@@ -509,6 +511,26 @@ def set_original_aten_op(func):
 
 
 
+# This mode is **only** used for pre_dispatch tracing.
+# In particular, we need to make sure that autograd/autocast API's
+# that do not desugar into dispatcher operators stay in the graph.
+class ProxyTorchFunctionMode(TorchFunctionMode):
+
+    def __init__(self, tracer):
+        self.tracer = tracer
+
+    def __torch_function__(self, func, types, args=(), kwargs=None):
+        kwargs = kwargs or {}
+        pre_dispatch_ops = [
+            torch._C._set_grad_enabled,
+            torch.amp._enter_autocast,
+            torch.amp._exit_autocast,
+        ]
+        if func in pre_dispatch_ops:
+            return self.tracer.create_node("call_function", func, args, {})
+            # Don't actually run the function! We just want to trace the calls
+            # into a graph. We don't actualy want to change global autograd state.
```

Comment:
This would be wrong if you weren't doing pre-autograd/pre-autocast tracing, right? Because if you are, you DO have to actually run these to make sure state is appropriately updated

Reply:
Yes - is the argument here that we should be clearer about the fact that this mode is only valid to use when we're doing pre-dispatch tracing? (I'll update the name to `PreDispatchTorchProxyFunctionMode` at least, but open to suggestions)